src/porepy/numerics/interface_laws/contact_mechanics_interface_laws.py

```diff
@@ -0,0 +1,781 @@
+        # Stress contribution from boundary conditions.
+        rhs[mortar_ind] = -(
+            mg.master_to_mortar_int(nd=ambient_dimension)
+            * sign_switcher
+            * master_bound_stress
+            * master_bc_values
+        )
+        # The stress contribution from the mortar variables, mapped to the higher
+        # dimensional domain via a boundary condition, and back again by a
+        # projection operator.
+        # Switch the direction of the vectors, so that for all faces, a positive
+        # force points into the fracture surface.
+        stress_from_mortar = (
+            mg.master_to_mortar_int(nd=ambient_dimension)
+            * sign_switcher
+            * master_bound_stress
+            * mg.mortar_to_master_avg(nd=ambient_dimension)
+        )
+        cc[mortar_ind, mortar_ind] = stress_from_mortar
+
+        ## Second, the contact stress is mapped to the mortar grid.
+        # We have for the positive (first) and negative (second) side of the mortar that
+        # \lambda_slave = \lambda_mortar_pos = -\lambda_mortar_neg,
+        # so we need to map the slave traction with the corresponding signs to match the
+        # mortar tractions.
+
+        # The contact forces are defined in the surface coordinate system.
+        # Map to the mortar grid, and rotate back again to the global coordinates
+        # (note the inverse rotation is given by a transpose).
+        # Finally, the contact stresses will be felt in different directions by
+        # the two sides of the mortar grids (Newton's third law), hence
+        # adjust the signs
+        contact_stress_to_mortar = (
+            mg.sign_of_mortar_sides(nd=ambient_dimension)
+            * projection.project_tangential_normal(mg.num_cells).T
+            * mg.slave_to_mortar_int(nd=ambient_dimension)
+        )
+        # Minus to obtain -\lambda_slave + \lambda_mortar = 0.
+        cc[mortar_ind, slave_ind] = -contact_stress_to_mortar
+
+        if self.use_surface_discr:
+            restrict_to_tangential_direction = projection.project_tangential(
+                mg.num_cells
+            )
+
+            # The first block contains the surface diffusion component. This has
+            # the surface diffusion operator for the mortar variables, and a
+            # mapping of contact forces on the slave variables.
+            # The second block gives continuity of forces in the normal direction.
+            surface_discr = matrix_dictionary_edge[self.SURFACE_DISCRETIZATION_KEY]
+
+            cc[mortar_ind, mortar_ind] += (
+                restrict_to_tangential_direction.T * surface_discr
+            )
+
+        matrix += cc
+
+        return matrix, rhs
+
+
+class MatrixScalarToForceBalance:
+    """
+    This class adds the matrix scalar (pressure) contribution to the force balance posed
+    on the mortar grid by PrimalContactCoupling.
+
+    We account for the grad P contribution to the forces on the higher-dimensional
```

Comment:
It is not really the gradient of the pressure, just the pressure

Reply:
yes